icu4c/source/common/uresbund.cpp

```diff
@@ -2026,7 +2026,7 @@ ures_getValueWithFallback(const UResourceBundle *bundle, const char *path,
         }
     }
     value.setData(&rb->fResData);
-    value.setResource(rb->fRes);
+    value.setResource(rb->fRes, ResourceTracer(rb));
```

Comment:
I think this doesn't actually work.  The `ResourceTracer` is stored by value in the `ResourceDataValue` and retains a pointer to the `rb`.  It then looks inside the `rb` to figure out the full path only when a data access method is used, which happens after `rb` leaves scope.

I remember thinking about the `UResourceBundle` lifetime when writing `ResourceTracer` and concluded it was safe to retain a pointer because, up until this function, the base `rb` was always alive at least as long as the sugar classes (ResourceDataValue, etc.) used to peek inside of it.  The tracing walks up the tree of sugar classes until it finds a "real" `rb`, the ancestor from which data is being read.

Can you test it in valgrind/asan?  Build ICU with tracing enabled as described in the Markdown user guide, and run a quick test with LocaleMatcher.

If it doesn't work, you can fix it by pulling the `StackUResourceBundle` into the parent scope, as suggested before, either by pulling logic out of the function, or by passing in a reference to be populated.

Reply:
Note: the Travis ASAN test, on which the tracing code is enabled, is failing with the error I was expecting.  (Good catch from the build bots if I hadn't noticed this from code review!)

Reply:
Darn. When I wrote this function, there was no tracing yet, and I didn't want to have the caller know about the temp resource bundle because, unlike other functions you mentioned, the UResourceBundle is not a bona fide output of this function.

I can't think of another good, quick workaround, so I added the temp fill-in parameter now.

Reply:
With the new fill-in, would you also undo the hack where you modified ResourceDataValue to store the resData by value instead of by pointer?  Then we can keep ResourceDataValue as a lightweight helper class for accessing data, which retains pointers to the heavier objects like UResourceBundle that actually store the data.

Reply:
I don't want to put the pointer back into the ResourceDataValue until we do the real cleanup and change the low-level code so that there is only one copy of the resData per file, shared among all of its item UResourceBundle's which then should also have pointers to the resData, not value copies.